### DISCUSSION TITLE
Can we deliver bulk email using single connection?

### DISCUSSION THREAD

**Discussion by siddiquiaffan** [eleith/emailjs#331] [questions]
Can we deliver bulk email using single connection? Like in nodemailer there's an option for ` pool `

**Comment by zackschuster** [comment]
not with a simple toggle, but we do FIFO queueing on repeated calls to `send`/`sendAsync` & after the queue is empty we leave the connection open for one second in case a new message comes in. 

this should be enough to allow you to trivially implement some brand of pooling using promises & for-of loops, but you can let me know if you encounter difficulties doing so 😄

**Comment by siddiquiaffan** [comment]
> not with a simple toggle, but we do FIFO queueing on repeated calls to `send`/`sendAsync` & after the queue is empty we leave the connection open for one second in case a new message comes in.
> 
> this should be enough to allow you to trivially implement some brand of pooling using promises & for-of loops, but you can let me know if you encounter difficulties doing so 😄

Are there any docs for that? It'll be helpful for me.

**Comment by zackschuster** [comment]
it's more of a snippet kind of thing:

```typescript
import { Message, SMTPClient } from 'emailjs';

const client = new SMTPClient({ /* your configuration */ });
const messages: Message[] = [ /* your messages */ ];

// to queue one at a time:
for (const message of messages) {
	const response = await client.sendAsync(message);
}
// --- OR ---
// to queue all at once:
const responses = await Promise.allSettled(messages.map(message => client.sendAsync(message)));

**Comment by siddiquiaffan** [comment]
> it's more of a snippet kind of thing:
> 
> ```ts
> import { Message, SMTPClient } from 'emailjs';
> 
> const client = new SMTPClient({ /* your configuration */ });
> const messages: Message[] = [ /* your messages */ ];
> 
> // to queue one at a time:
> for (const message of messages) {
> 	const response = await client.sendAsync(message);
> }
> // --- OR ---
> // to queue all at once:
> const responses = await Promise.allSettled(messages.map(message => client.sendAsync(message)));
> ```

Are you sure by using this method all messages will be sent in one connection?

**Comment by zackschuster** [comment]
as long as less than one second passes between calls to `send`/`sendAsync`, it should re-use the same connection.

**Comment by siddiquiaffan** [comment]
> as long as less than one second passes between calls to `send`/`sendAsync`, it should re-use the same connection.

Ohhkay will check it out. Thanks alot!

**Comment by trasherdk** [comment]
Maybe this: https://github.com/trasherdk/emailjs/issues/4
